### PR TITLE
feat: typed route generation for expo router

### PR DESCRIPTION
**Pull request by marklawlor**
> Draft PR - just seeking feedback on the implementation.

# Why

WIP implementation of route generation for Expo Router.

# How

## Typescript bootstrapping

This is expands the Typescript bootstrapping to also accommodate for long-lived typescript watchers.
- Changes `waitForMetroToObserveTypeScriptFile` to `metroWatchTypeScriptFiles`
- Processes are now responsible to disabling their watcher
- tsconfig can optionally be watched

## Type generation

`typedRouteGenerator` is invokes after the CLI has bootstrapped TypeScript.
- Generates types only for absolute paths
- Supports static and dynamic routes
- Supports URL normalisation (TODO: Need unit tests for these!)
- Supports dynamic route parameter typing

# Test Plan

`EXPO_ROUTER_TYPED_ROUTES=true nexpo start`

// Todo

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by marklawlor** [comment]
- [ ] Need to check for existence of expo router

File: packages/@expo/cli/src/start/server/type-generation/index.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  metro: Server | null;
+  projectRoot: string;
+}
+
+export async function typescriptTypeGeneration({
+  metro,
+  projectRoot,
+  server,
+}: TypeScriptTypeGenerationOptions) {
+  if (!env.EXPO_ROUTER_TYPED_ROUTES) {
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/type-generation/index.ts:22]:
```suggestion
  if (!env.EXPO_USE_TYPED_ROUTES) {
```

`EXPO_USE_` is our standard for enabling a feature, and `EXPO_NO_` is used for disabling.

File: packages/@expo/cli/src/start/server/type-generation/routes.ts [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+  for (const group of groups) {
+    const groupRoute = route.replace(groupsMatch, `/(${group})/`);
+    expandGroupRoutes(groupRoute, routes);
+  }
+
+  return routes;
+}
+
+const routerDotTSTemplate = unsafeTemplate`declare module "expo-router" {
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/type-generation/routes.ts:185]:
```suggestion
// NOTE: This code refers to a specific version of `expo-router` and is therefore unsafe to
// mix with arbitrary versions.
// TODO: Version this code with `expo-router` or version expo-router with `@expo/cli`.
const routerDotTSTemplate = unsafeTemplate`declare module "expo-router" {
```

File: packages/@expo/cli/src/utils/env.ts [hunk 3]
```diff
@@ -153,6 +153,14 @@ class Env {
   get EXPO_USE_CUSTOM_INSPECTOR_PROXY(): boolean {
     return boolish('EXPO_USE_CUSTOM_INSPECTOR_PROXY', false);
   }
+
+  /** Enable Expo Router typed router generation */
+  get EXPO_ROUTER_TYPED_ROUTES() {
+    return boolish('EXPO_ROUTER_TYPED_ROUTES', false);
+  }
+  get EXPO_ROUTER_APP_ROOT() {
+    return string('EXPO_ROUTER_APP_ROOT', '../../app');
```
Comment (EvanBacon) [inline packages/@expo/cli/src/utils/env.ts:162]: This cannot have a default value, it should always be calculated to ensure the root is correctly relative to the source entry.

File: packages/@expo/cli/src/start/server/DevServerManager.ts [hunk 4]
```diff
@@ -152,16 +153,26 @@ export class DevServerManager {
...
     // Optionally, wait for the user to add TypeScript during the
     // development cycle.
     const server = devServers.find((server) => server.name === 'metro');
     if (!server) {
       return;
     }
-    await server.waitForTypeScriptAsync();
+
+    if (!typescriptPrerequisite) {
+      server.waitForTypeScriptAsync();
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/DevServerManager.ts:172]: Does this mean the user needs to restart the process to enable route generation?

File: packages/@expo/cli/src/start/server/type-generation/routes.ts [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+export const CAPTURE_GROUP_REGEX = /[\\(,](\w+?)(?=[,\\)])/g;
+
+export interface SetupTypedRoutesOptions {
+  server: ServerLike;
+  metro: Server;
+  typesDirectory: string;
+}
+
+export async function setupTypedRoutes({ server, metro, typesDirectory }: SetupTypedRoutesOptions) {
+  const appRoot = path.resolve(env.EXPO_ROUTER_APP_ROOT);
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/type-generation/routes.ts:30]: This value is relative to the `node_module/expo-router/entry` file, so it cannot be resolved like this. The actual resolution is more like this https://github.com/expo/expo/blob/036e9444308bd944b04f0d4537f0c055e344b8d0/packages/%40expo/cli/src/start/server/metro/router.ts#LL18C20-L18C51

Can add a function to that file like `getExpoRouterRootDirectory(projectRoot: string): string` which just returns `path.join(projectRoot, 'app')` then in the future we can extend it to support custom directories.

File: packages/@expo/cli/src/utils/env.ts [hunk 6]
```diff
@@ -153,6 +153,14 @@ class Env {
   get EXPO_USE_CUSTOM_INSPECTOR_PROXY(): boolean {
     return boolish('EXPO_USE_CUSTOM_INSPECTOR_PROXY', false);
   }
+
+  /** Enable Expo Router typed router generation */
+  get EXPO_USE_TYPED_ROUTES() {
```
Comment (EvanBacon) [inline packages/@expo/cli/src/utils/env.ts:158]: Add to docs https://github.com/expo/expo/blob/main/docs/pages/workflow/expo-cli.mdx#environment-variables

File: packages/@expo/cli/src/utils/env.ts [hunk 7]
```diff
@@ -153,6 +153,14 @@ class Env {
   get EXPO_USE_CUSTOM_INSPECTOR_PROXY(): boolean {
     return boolish('EXPO_USE_CUSTOM_INSPECTOR_PROXY', false);
   }
+
+  /** Enable Expo Router typed router generation */
```
Comment (EvanBacon) [inline packages/@expo/cli/src/utils/env.ts:157]:
```suggestion
  /** **Experimental:** Enable automatic TypeScript types for Expo Router projects (SDK +49). */
```

File: packages/@expo/cli/src/start/server/type-generation/tsconfig.ts [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    updates.add('include');
+  } else {
+    if (!tsConfig.include.includes('.expo/types/**/*.ts')) {
+      tsConfig.include = [...tsConfig.include, '.expo/types/**/*.ts'];
+      updates.add('include');
+    }
+  }
+
+  if (updates.size) {
+    await JsonFile.writeAsync(tsConfigPath, tsConfig);
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/type-generation/tsconfig.ts:30]: Add clean up when the feature isn't enabled to ensure stale types aren't being used.

Comment (EvanBacon) [reply packages/@expo/cli/src/start/server/type-generation/tsconfig.ts:30]: Also probably worth testing this function.

File: packages/@expo/cli/src/utils/env.ts [hunk 9]
```diff
@@ -153,6 +153,14 @@ class Env {
...
     return boolish('EXPO_USE_CUSTOM_INSPECTOR_PROXY', false);
   }
+
+  /** **Experimental:** Enable automatic TypeScript types for Expo Router projects (SDK +49). */
+  get EXPO_USE_TYPED_ROUTES() {
+    return boolish('EXPO_USE_TYPED_ROUTES', false);
+  }
+  get EXPO_ROUTER_APP_ROOT() {
+    return string('EXPO_ROUTER_APP_ROOT');
+  }
```
Comment (EvanBacon) [inline packages/@expo/cli/src/utils/env.ts:163]:
```suggestion
```

File: packages/@expo/cli/src/start/server/type-generation/utils.ts [hunk 10]
```diff
@@ -0,0 +1,8 @@
+/**
+ * Convert a set to a string type.
+ * @example setToType(new Set(['a', 'b'])) => 'a | b'
+ * @example setToType() => 'never'
+ */
+export const setToUnionType = <T>(set: Set<T>) => {
```
Comment (EvanBacon) [inline packages/@expo/cli/src/start/server/type-generation/utils.ts:6]: Move to `src/utils/setToUnionType.ts`

**Review by EvanBacon — APPROVED** [review]

**Pull request merged**